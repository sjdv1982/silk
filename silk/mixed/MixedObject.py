from collections.abc import MutableMapping, MutableSequence
from . import MixedBase

class MixedObject(MixedBase, MutableMapping, MutableSequence):
    def _proxy(self):
        proxy = self._monitor.get_path(self._path)
        if isinstance(proxy, MixedObject):
            raise AttributeError(self._path)
        return proxy
    def _proxy2(self, item):
        proxy = self._monitor.get_path(self._path)
        if isinstance(proxy, MixedObject):
            if isinstance(item, int):
                self._monitor.set_path(self._path, [])
            elif isinstance(item, str):
                self._monitor.set_path(self._path, {})
            else:
                raise TypeError(item)
            proxy = self._proxy()
        return proxy
    def __getitem__(self, item):
        try:
            proxy = self._proxy2(item)
        except AttributeError:
            raise KeyError(item)
        return proxy.__getitem__(item)
    def __setitem__(self, item, value):
        proxy = self._proxy2(item)
        return proxy.__setitem__(item, value)
    def __delitem__(self, item):
        proxy = self._proxy()
        return proxy.__delitem__(item)
    def __iter__(self):
        proxy = self._proxy()
        return proxy.__iter__()
    def __len__(self):
        proxy = self._proxy()
        return proxy.__len__()
    def insert(self, item, value):
        proxy = self._proxy()
        return proxy.insert(item, value)
    def append(self, value):
        proxy = self._monitor.get_path(self._path)
        if isinstance(proxy, MixedObject):
            self._monitor.set_path(self._path, [])
            proxy = self._proxy()
        return proxy.append(value)
    def update(self, value):
        proxy = self._monitor.get_path(self._path)
        if isinstance(proxy, MixedObject):
            self._monitor.set_path(self._path, {})
            proxy = self._proxy()
        return proxy.update(value)
    def clear(self):
        proxy = self._monitor.get_path(self._path)
        if isinstance(proxy, MixedObject):
            self._monitor.set_path(self._path, {})
            proxy = self._proxy()
        return proxy.clear()
