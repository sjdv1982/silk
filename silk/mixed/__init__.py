"""
Python types inside Numpy arrays must be containers, strings, or None, not scalars.
Typedefs are either a string or a typedict with a "type" field containing such a string.
Type strings can be JSON schema types or "tuple", in which case "shape" must be defined
An "array" or "tuple" type contains a field "items", and a field "identical". If "identical" is True, then "items" contains only a single type/typedict,
"object" typedef always contain a field "properties"
A NumPy array (or tuple) of Python objects also has a field "item_storage", which can be "pure-plain" or "mixed-plain", depending on the Python type
If "identical" is False, "item_storage" is a list
In a typedict, a child's "storage" is stored by the parent into a child typedict, but only if:
   1) the child is not pure; or 2) the child is binary and the parent is plain (or vice versa)
"""

MAGIC_NUMPY = b"\x93NUMPY"
MAGIC_SEAMLESS = b'\x93SEAMLESS'
MAGIC_SEAMLESS_MIXED = b'\x94SEAMLESS-MIXED'

import numpy as np
from functools import partialmethod

from ..SilkBase import binary_special_method_names
from .. import Wrapper
from ..validation import (
  _array_types, _integer_types, _float_types, _string_types, _unsigned_types,
  _allowed_types, Scalar, is_np_struct, FormWrapper
)

scalars = ("boolean", "integer", "number", "string")

import numpy as np

def is_numpy_buffer(buffer):
    if buffer is None:
        return False
    return buffer[:len(MAGIC_NUMPY)] == MAGIC_NUMPY

class MonitorTypeError(TypeError):
    pass

class MixedBase(Wrapper):
    def __init__(self, _monitor, _path):
        self._monitor = _monitor
        self._path = _path
    def _unwrap(self):
        return FormWrapper(self.value, self.form, self.storage)
    @property
    def value(self):
        data = self._monitor.get_data(self._path)
        return data
    @property
    def form(self):
        return self._monitor.get_form(self._path)
    @property
    def storage(self):
        return self._monitor.get_storage(self._path)
    def set(self, value):
        self._monitor.set_path(self._path, value)
    def __setattr__(self, attr, value):
        if attr.startswith("_"):
            return super().__setattr__(attr, value)
        return self.__setitem__(attr, value)
        raise AttributeError(attr)
    def __getattr__(self, attr):
        if attr.startswith("_"):
            raise AttributeError(attr)
        if isinstance(self.value, np.ndarray):
            return getattr(self.value, attr)
        else:
            return self.__getitem__(attr)
        raise AttributeError(attr)
    def __str__(self):
        return str(self.value)
    def __repr__(self):
        return repr(self.value)

def mixed_scalar_binary_method(self, other, name):
    if isinstance(other, MixedBase):
        other = other.value
    return getattr(self.value,name)(other)

def mixed_scalar_binary_method_inplace(self, other, name2):
    result = getattr(self.value,name2)(other)
    if result is NotImplemented:
        return NotImplemented
    self._monitor.set_path(self._path, result)
    return result

class MixedScalar(MixedBase):
    def __getitem__(self, item):
        value = self.value
        if not isinstance(value, (str, np.ndarray)):
            raise TypeError(type(value))
        return value[item]
    def __getattr__(self, attr):
        if attr.startswith("_"):
            raise AttributeError(attr)
        if isinstance(self.value, np.ndarray):
            return getattr(self.value, attr)
        elif isinstance(attr, str):
            try:
                return self.__getitem__(attr)
            except TypeError:
                raise AttributeError(attr) from None
        raise AttributeError(attr)


for name in binary_special_method_names:
    if name.startswith("__i"):
        name2 = "__" + name[3:]
        m = partialmethod(mixed_scalar_binary_method_inplace, name2=name2)
    else:
        m = partialmethod(mixed_scalar_binary_method, name=name)
    setattr(MixedScalar, name, m)


from .MixedDict import MixedDict
from .MixedList import MixedList
from .MixedObject import MixedObject
from .Monitor import Monitor
from .Backend import Backend, DefaultBackend, SilkBackend
from .get_form import is_contiguous, is_unsigned
