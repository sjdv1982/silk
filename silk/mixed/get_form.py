"""
Paradigm:
- get_form_XXX as long as the data is in plain storage mode
- When we encounter a numpy array Q, we go into binary storage mode:
    - First determine the form based on dtype: get_tform_XXX
    - Skip Python objects inside the dtype (storage = None)
    - Then, visit (= fill in) those Python objects using the data: visit_typedef_XXX
    - Re-enter plain mode as soon as we encounter these Python objects: get_form_XXX
    - After each Python object, leave plain-mode and return to binary
    - After finishing the Numpy array Q, leave binary mode and return to plain
"""
import numpy as np
from numpy import ndarray, void
from copy import deepcopy
from . import ( Scalar,
  _array_types, _integer_types, _float_types, _string_types, _unsigned_types
)
_string_types = (str,)  ##JSON cannot deal with bytes

dt_builtins = (
    np.dtype("bool"),
    np.dtype("int8"),
    np.dtype("int16"),
    np.dtype("int32"),
    np.dtype("int64"),
    np.dtype("uint8"),
    np.dtype("uint16"),
    np.dtype("uint32"),
    np.dtype("uint64"),
    np.dtype("float32"),
    np.dtype("float64"),
)
def is_contiguous(data):
    """Returns if the data is C-contiguous
    The last stride must be itemsize (bytesize),
     the second-last must be itemsize * shape[-1], etc.
    These criteria are stricter than Numpy's c_contiguous:
      see the documentation of numpy.ndarray.flags for an explanation
    """
    if not data.flags["C_CONTIGUOUS"]:
        return False
    stride = data.itemsize
    contiguous = False
    for n in range(data.ndim-1, -1, -1):
        if data.strides[n] != stride:
            break
        stride *= data.shape[n]
    else:
        contiguous = True
    return contiguous

def is_unsigned(dt):
    return any([dt == t for t in _unsigned_types])

def get_typedef_scalar(value):
    if isinstance(value, bool):
        typedef = "boolean"
    elif isinstance(value, _integer_types):
        typedef = "integer"
    elif isinstance(value, _float_types):
        typedef = "number"
    elif isinstance(value, _string_types):
        typedef = "string"
    elif value is None:
        typedef = "null"
    else:
        raise TypeError(type(value))
    return typedef


def visit_typedef_numpy_items(item_typedef, data):
    #numpy items of mixed-binary storage
    identical = True
    items = item_typedef
    items2 = []
    for d in data:
        item_typedef_curr = deepcopy(item_typedef)
        _ = visit_typedef_numpy("mixed-binary", item_typedef_curr, d)
        items2.append(item_typedef_curr)
    items = items2[0]
    for item_typedef_curr in items2[1:]:
        if item_typedef_curr != items2[0]:
            identical = False
            items = items2
            break
    return items, identical

def visit_typedef_numpy_tuple(storage0, typedef, data):
    # An Numpy tuple field of mixed-binary
    assert storage0 == "mixed-binary" #If None, call visit_typedef_numpy instead
    assert typedef["identical"]
    assert typedef["type"] == "tuple"

    assert data.shape == typedef["shape"]
    item_typedef = typedef["items"]
    items, identical = visit_typedef_numpy_items(item_typedef, data)
    typedef["identical"] = identical
    typedef["item_storage"] = item_storage
    typedef["items"] = item_typedef
    return "mixed-binary"

def visit_typedef_numpy_struct(storage0, typedef, data):
    assert "properties" in typedef ##since type is "object"
    props = typedef["properties"]
    storage = "pure-binary"
    for k in props:
        subtypedef = props[k]
        if not "storage" in subtypedef:
            continue
        subdata = data[k]
        substorage0 = subtypedef["storage"]
        substorage = visit_typedef_numpy(substorage0, subtypedef, subdata)
        if substorage0 is None or substorage != "pure-binary":
            storage = "mixed-binary"
        subtypedef["storage"] = substorage
    if storage == "mixed-binary":
        for k in props:
            subtypedef = props[k]
            substorage = subtypedef.get("storage")
            if substorage is None:
                subtypedef["storage"] = "pure-binary"
            elif substorage == "mixed-binary":
                subtypedef.pop("storage")
    return storage

def visit_typedef_numpy(storage0, typedef, data):
    #visit and fill in mixed-binary typedefs
    if storage0 is None: #typedef of a Python object inside Numpy
        assert not isinstance(data, void), type(data)
        if isinstance(typedef, dict) and "items" in typedef:
            assert typedef["items"] is None
            storage = "mixed-binary"
            if typedef["type"] in ("array", "tuple"): # Numpy array of Python objects
                if typedef["type"] == "tuple": # Numpy multi-element field of Python objects
                    assert data.shape == typedef["shape"]
                item_storage, item_typedef = [], []
                for d in data: #TODO: multi-dimensional (currently NotImplementedError)
                    curr_item_storage, curr_item_typedef = get_form_python_inside_numpy(d)
                    item_storage.append(curr_item_storage)
                    item_typedef.append(curr_item_typedef)
                for dnr, d in enumerate(data[1:]):
                    if item_storage[dnr] != item_storage[0] or \
                      item_typedef[dnr] != item_typedef[0]:
                        identical = False
                        break
                else:
                    identical = True
                    item_storage = item_storage[0]
                    item_typedef = item_typedef[0]
                typedef["identical"] = identical
                typedef["item_storage"] = item_storage
                typedef["items"] = item_typedef
            else:
                raise TypeError(typedef["type"])
        else: # Python object inside a Numpy structured dtype
            if isinstance(data, ndarray):
                storage, typedef2 = get_form_list(data)  #plain storage
            else:
                storage, typedef2 = get_form_python_inside_numpy(data)  #plain storage
            if isinstance(typedef2, dict):
                typedef.clear()
                typedef.update(typedef2)
            else:
                typedef = typedef2
    elif storage0 == "mixed-binary":
        assert isinstance(data, (ndarray, void)), type(data)
        assert isinstance(typedef, dict), type(typedef)
        if typedef["type"] == "object":
            return visit_typedef_numpy_struct(storage0, typedef, data)
        elif typedef["type"] == "array":
            raise Exception #should be impossible: arbitrary-length Numpy arrays inside Numpy dtypes are Python objects (storage0 = None)!
        elif typedef["type"] == "tuple":
            return visit_typedef_numpy_tuple(storage0, typedef, data)
        else:
            raise TypeError(typedef["type"])
    else:
        raise ValueError(storage0)
    return storage

def get_form_python_inside_numpy(data):
    if isinstance(data, list):
        return get_form_list_plain(data)
    elif isinstance(data, dict):
        return get_form_dict_plain(data)
    elif isinstance(data, _string_types):
        return "pure-plain", "string"
    elif data is None:
        return "pure-plain", "null"
    else:
        raise TypeError(type(data)) #must be list, dict, string or None; scalar Python objects not allowed

def get_tform_numpy_pyobject(dt):
    if dt.ndim:
        if dt.ndim > 1: raise NotImplementedError #TODO: multi-dimensional Python tuples
        storage = None #technically, mixed-binary, but this will be visited
        typedef = {
            "type": "tuple",
            "items": None, #to be visited
            "shape": dt.shape,
        }
    else:
        storage, typedef = None, None #to be visited
    return storage, typedef

def get_tform_numpy_builtin(dt):
    dtb = dt.base
    if dtb == object:
        return get_tform_numpy_pyobject(dt)

    storage = "pure-binary"
    if dtb == bool:
        typedef0 = "boolean"
    elif any([dtb == t for t in _integer_types]):
        typedef0 = "integer"
    elif any([dtb == t for t in _float_types]):
        typedef0 = "number"
    elif is_np_str(dtb):
        typedef0 = "string"
    else:
        raise TypeError(dtb)

    typedef = {
        "type": typedef0,
        "bytesize": dt.base.itemsize,
    }
    if typedef0 == "integer":
        unsigned = is_unsigned(dt)
        typedef["unsigned"] = unsigned
    if dt.ndim:
        typedef = {
            "type": "tuple",
            "items": typedef,
            "identical": True,
            "shape": dt.shape,
        }
    return storage, typedef

def get_tform_numpy_struct(dt):
    if not dt.isalignedstruct:
        raise TypeError("Composite dtypes must be memory-aligned", dt)
    if not dt.isnative:
        raise TypeError("Composite dtypes must be native")
    storages = {}
    props = {}
    typedef = {"type": "object", "properties": props, "bytesize": dt.itemsize}
    typedef["order"] = list(dt.fields)
    for fieldname in dt.fields:
        cstorage, ctypedef = get_tform_numpy(dt[fieldname])
        storages[fieldname] = cstorage
        props[fieldname] = ctypedef
    storage_set = set(storages.values())
    if len(storage_set) == 1 and storage_set.pop() == "pure-binary":
        storage = "pure-binary"
    else:
        storage = "mixed-binary"
        for fieldname in dt.fields:
            cstorage = storages[fieldname]
            if cstorage == "pure-binary":
                continue
            ctypedef = props[fieldname]
            if ctypedef is None or isinstance(ctypedef, str):
                ctypedef = {"type": ctypedef}
                props[fieldname] = ctypedef
            ctypedef["storage"] = cstorage
    if dt.ndim:
        typedef["storage"] = storage
        typedef = {
            "type": "tuple",
            "shape": dt.shape,
            "items":  typedef,
            "identical": True
        }
    return storage, typedef


def is_np_str(dt):
    return dt == np.dtype("S%d" % dt.itemsize) or dt == np.dtype("U%d" % (dt.itemsize/4))

def get_tform_numpy(dt):
    if dt.base.isbuiltin or is_np_str(dt) or dt in dt_builtins:
        return get_tform_numpy_builtin(dt)
    return get_tform_numpy_struct(dt)

def get_form_dict_plain(data):
    typedef = {"type": "object"}
    if not len(data):
        return "pure-plain", typedef
    props = {}
    typedef["properties"] = props
    storages = {}
    for k,v in data.items():
        cstorage, ctypedef = get_form(v)
        props[k] = ctypedef
        storages[k] = cstorage
    storage_set = set(storages.values())
    if len(storage_set) == 1 and storage_set.pop() == "pure-plain":
        storage = "pure-plain"
    else:
        storage = "mixed-plain"
        for fieldname in data:
            cstorage = storages[fieldname]
            if cstorage == "pure-plain":
                continue
            ctypedef = props[fieldname]
            if isinstance(ctypedef, str):
                ctypedef = {"type": ctypedef}
                props[fieldname] = ctypedef
            ctypedef["storage"] = cstorage
    return storage, typedef

def get_form_dict(data):
    if isinstance(data, Scalar):
        raise TypeError
    elif isinstance(data, void):
        dt = data.dtype
        storage, typedef = get_tform_numpy(dt)
        if storage in (None, "mixed-binary"):
            storage = visit_typedef_numpy(storage, typedef, data)
    elif isinstance(data, _array_types):
        raise TypeError
    elif isinstance(data, dict):
        storage, typedef = get_form_dict_plain(data)
    else:
        raise TypeError
    return storage, typedef

def get_form_items_numpy(data):
    dt = data.dtype
    storage, item_typedef = get_tform_numpy(dt)
    identical = True
    items = item_typedef
    if storage == "pure-binary":
        pass
    elif storage == "mixed-binary":
        items, identical = visit_typedef_numpy_items(item_typedef, data)
    elif storage is None: #pyobject
        return None, item_typedef, True #Will be visited later
    else:
        raise ValueError(storage)
    return storage, items, identical

def get_form_items_list_plain(data):
    identical = True
    storages = []
    items2 = []
    for d in data:
        item_storage, item_typedef = get_form(d)
        storages.append(item_storage)
        items2.append(item_typedef)

    storage_set = set(storages)
    if len(storage_set) > 1:
        different = True
    else:
        different = False
        for item_typedef in items2[1:]:
            if item_typedef != items2[0]:
                different = True
                break
    if different:
        items = items2
        identical = False
        if len(storage_set) == 1 and storage_set.pop() == "pure-plain":
            storage = "pure-plain"
        else:
            storage = "mixed-plain"
        for n in range(len(data)):
            s = storages[n]
            if s == "pure-plain":
                continue
            ctypedef = items[n]
            if isinstance(ctypedef, str):
                ctypedef = {"type": ctypedef}
                items[n] = ctypedef
            items[n]["storage"] = s
    else:
        if not len(items2):
            storage = "pure-plain"
            items = None
        else:
            items = items2[0]
            child_storage = storages[0]
            if child_storage == "pure-plain":
                storage = "pure-plain"
            else:
                storage = "mixed-plain"
                if isinstance(items, str):
                    items = {"type": items}
                items["storage"] = child_storage
    return storage, items, identical

def get_form_list(data):
    extra = {}
    if isinstance(data, ndarray):
        dt = data.dtype
        if not dt.isnative:
            raise TypeError("dtypes must be native")
        storage, items, identical = get_form_items_numpy(data)
        extra = {
            "shape": data.shape,
        }
        strides = data.strides
        contiguous = is_contiguous(data)
        extra["contiguous"] = contiguous
        if not contiguous:
            extra["strides"] = strides
    elif isinstance(data, _array_types):
        storage, items, identical = get_form_items_list_plain(data)
        extra = {
            "shape": (len(data),),
        }
    elif isinstance(data, dict):
        raise TypeError
    elif isinstance(data, Scalar):
        raise TypeError
    elif isinstance(data, void):
        raise TypeError
    else:
        raise TypeError

    typedef = {
        "type": "array",
        "items": items,
        "identical": identical
    }
    typedef.update(extra)
    return storage, typedef

def get_form_list_plain(data):
    #data must be a plain list
    storage, items, identical = get_form_items_list_plain(data)
    typedef = {
        "type": "array",
        "items": items,
        "identical": identical,
        "shape": (len(data),),
    }
    return storage, typedef


def get_form(data):
    if isinstance(data, bytes):
        data = np.array(data)
    if isinstance(data, Scalar):
        storage, typedef = "pure-plain", get_typedef_scalar(data)
    elif isinstance(data, void):
        dt = data.dtype
        storage, typedef = get_tform_numpy(dt)
        if storage in (None, "mixed-binary"):
            storage = visit_typedef_numpy(storage, typedef, data)
    elif isinstance(data, _array_types):
        storage, typedef = get_form_list(data)
    elif isinstance(data, dict):
        storage, typedef = get_form_dict_plain(data)
    else:
        raise TypeError(type(data))
    assert storage in (None, "pure-plain", "mixed-plain", "pure-binary", "mixed-binary"), storage
    return storage, typedef
