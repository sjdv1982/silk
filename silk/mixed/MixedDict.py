from numpy import ndarray, void
from collections.abc import MutableMapping
from . import MixedBase, is_np_struct
from .get_form import get_form_dict

class MixedDict(MixedBase, MutableMapping):
    def __getitem__(self, item):
        path = self._path + (item,)
        try:
            return self._monitor.get_path(path)
        except TypeError:
            raise KeyError
    def __setitem__(self, item, value):
        path = self._path + (item,)
        self._monitor.set_path(path, value)
    def __delitem__(self, item):
        path = self._path + (item,)
        self._monitor.del_path(path)
    def __iter__(self):
        data = self._monitor.get_data(self._path)
        if data is None:
            data = {}
        return iter(data)
    def __len__(self):
        data = self._monitor.get_data(self._path)
        if data is None:
            data = {}
        return len(data)
    def clear(self):
        data = self._monitor.get_data(self._path)
        if data is not None:
            for path in list(data.keys()):
                if isinstance(path, str):
                    path = (path,)
                self._monitor.del_path(path)
    def update(self, other):
        if isinstance(other, MixedBase):
            other = other.value
        for k,v in other.items():
            path = self._path + (k,)
            self._monitor.set_path(path, v)


class MixedNumpyStruct(MixedBase, MutableMapping):
    def __getitem__(self, item):
        path = self._path + (item,)
        return self._monitor.get_path(path)
    def __setitem__(self, item, value):
        path = self._path + (item,)
        self._monitor.set_path(path, value)
    def __delitem__(self, item):
        raise TypeError("Cannot delete Numpy struct item '%s'" % item)
    def __iter__(self):
        data = self._monitor.get_data(self._path)
        return data.dtype.fields
    def __len__(self):
        data = self._monitor.get_data(self._path)
        return len(data.dtype.fields)
