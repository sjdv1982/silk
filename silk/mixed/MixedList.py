from numpy import ndarray, void
from collections.abc import MutableSequence
from . import MixedBase
from .get_form import get_form_list

class MixedList(MixedBase, MutableSequence):
    def __getitem__(self, item):
        path = self._path + (item,)
        return self._monitor.get_path(path)
    def __setitem__(self, item, value):
        path = self._path + (item,)
        self._monitor.set_path(path, value)
    def insert(self, item, value):
        path = self._path + (item,)
        self._monitor.insert_path(path, value)
    def append(self, value):
        item = len(self)
        path = self._path + (item,)
        ###self._monitor.insert_path(path, value)
        self._monitor.set_path(path, value)
    def __delitem__(self, item):
        path = self._path + (item,)
        self._monitor.del_path(path)
    def __iter__(self):
        data = self._monitor.get_data(self._path)
        if data is None:
            data = []
        return iter(data)
    def __len__(self):
        data = self._monitor.get_data(self._path)
        if data is None:
            data = []
        return len(data)

class MixedNumpyArray(MixedBase, MutableSequence):
    def __getitem__(self, item):        
        path = self._path + (item,)
        return self._monitor.get_path(path)
    def __setitem__(self, item, value):
        path = self._path + (item,)
        self._monitor.set_path(path, value)
    def insert(self, item, value):
        raise TypeError("Cannot insert Numpy array item '%d'" % item)
    def __delitem__(self, item):
        raise TypeError("Cannot delete Numpy array item '%d'" % item)
    def __iter__(self):
        data = self._monitor.get_data(self._path)
        return iter(data)
    def __len__(self):
        data = self._monitor.get_data(self._path)
        return len(data)

