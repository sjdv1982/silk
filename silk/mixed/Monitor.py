import numpy as np
from numpy import ndarray, void
from .get_form import get_form
from . import MixedScalar, MixedBase, Scalar,  scalars, is_np_struct, _allowed_types
from . import MonitorTypeError
from .Backend import Backend
import json
from copy import deepcopy
from .. import Wrapper
from numpy import ndarray, void

class Monitor:
    def __init__(self, backend):
        assert isinstance(backend, Backend)
        self.backend = backend

    def get_instance(self, subform, path):
        if subform is None:
            if self.backend.formless:
                dataclass = None
                if len(path) == 0:
                    dataclass = self.backend.toplevel_dataclass
                if dataclass is None:
                    data = self._get_data(path)
                    dataclass = type(data)
                if dataclass is None:
                    raise KeyError(path)
                if issubclass(dataclass, dict):
                    return MixedDict(self, path)
                elif issubclass(dataclass, list):
                    return MixedList(self, path)
                elif issubclass(dataclass, void):
                    return MixedNumpyStruct(self, path)
                elif issubclass(dataclass, ndarray):
                    return MixedNumpyArray(self, path)
                else:
                    return MixedScalar(self, path)
            else:
                if not len(path):
                    return MixedObject(self, path)
                raise KeyError(path)
        if isinstance(subform, str):
            type_ = subform
        else:
            type_ = subform["type"]
        if type_ == "object":
            if subform.get("storage","pure-plain").endswith("binary"):
                return MixedNumpyStruct(self, path)
            else:
                return MixedDict(self, path)
        elif type_ == "array":
            if subform.get("storage","pure-plain").endswith("binary"):
                return MixedNumpyArray(self, path) #ndarray has an immutable type
            else:
                return MixedList(self, path)
        elif type_ in scalars:
            return MixedScalar(self, path) #scalars are all immutable
        else:
            raise TypeError(type_)

    def get_path(self, path=()):
        subform = self._get_form(path)
        return self.get_instance(subform, path)

    def _get_data(self, path):
        subdata = self.backend.get_path(path)
        return subdata

    def get_data(self, path=()):
        return self._get_data(path)

    def _get_form(self, path):
        subform = self.backend.get_subform(path)
        return subform

    def get_form(self, path=()):
        return self._get_form(path)

    def get_storage(self, path=()):
        if not len(path):
            return self.backend.get_storage()
        if len(path) == 1:
            parent_storage = self.backend.get_storage()
            if parent_storage is None:
                if not self.backend.formless:
                    raise AttributeError(path)
        else:
            parent_subdata = self._get_data(path[:-1])
            parent_subform = self._get_form(path[:-1])
            if parent_subform is None:
                if not self.backend.formless:
                    raise AttributeError(path)
                else:
                    parent_storage = None
            else:
                parent_storage = parent_subform.get("storage")
                if parent_storage is None:
                    if isinstance(parent_subdata, (void, ndarray)):
                        parent_storage = "pure-binary"
                    else:
                        parent_storage = "pure-plain"
        subdata = self._get_data(path)
        subform = self._get_form(path)
        storage = None
        if isinstance(subform, dict):
            storage = subform.get("storage")
        if storage is None:
            if parent_storage is None:
                return None
            if parent_storage.endswith("binary"):
                return "pure-binary"
            else:
                return "pure-plain"
        else:
            return storage

    def set_path(self, path, subdata):
        if isinstance(subdata, MixedBase):
            subdata = subdata.value
        if not isinstance(subdata, _allowed_types):
            raise TypeError(type(subdata))
        if self.backend.plain:
            json.dumps(subdata)
        elif isinstance(subdata, bytes):
            subdata = np.array(subdata)
        self.backend.set_path(path, subdata)

    def insert_path(self, path, subdata):
        """
        Inserts subdata right before the insertion point "path"
        The insertion point must be a list item
        """
        if not isinstance(subdata, _allowed_types):
            raise TypeError(type(subdata))
        if not len(path):
            raise TypeError
        if not isinstance(path[-1], int):
            raise TypeError(path)
        if self.backend.plain:
            json.dumps(subdata)
        elif isinstance(subdata, bytes):
            subdata = np.array(subdata)
        self.backend.insert_path(path, subdata)

    def del_path(self, path):
        self.backend.del_path(path)

    def __str__(self):
        return str(self.get_path())

    def __repr__(self):
        return repr(self.get_path())

from .MixedObject import MixedObject
from .MixedDict import MixedDict, MixedNumpyStruct
from .MixedList import MixedList, MixedNumpyArray
