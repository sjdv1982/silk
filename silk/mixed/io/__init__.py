from ..get_form import get_form
from .to_stream import to_stream
from .from_stream import from_stream
from .serialization import serialize, deserialize