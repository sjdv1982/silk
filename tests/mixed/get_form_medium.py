#medium-complexity tests for mixed get_form

import sys
import numpy as np
from silk.mixed.get_form import get_form
import json

_print = print
def print(*args):
    for a in args:
        _print(json.dumps(a, indent=2, sort_keys=True), end=" ")
    _print()


dt = np.dtype([
    ("a", int),
    ("b", object),
], align=True)
data = np.zeros(2, dtype=dt)
data[0]["b"] = [1,2,3]
data[1]["b"] = [4,5,6]
storage, form = get_form(data)
print(storage, form); print()

data[1]["b"] = [4.0,5.0,6.0]
storage, form = get_form(data)
print(storage, form); print()

arr = np.zeros(10)

data = [arr, arr, arr]
storage, form = get_form(data)
print(storage, form); print()

data = [5, arr, None]
storage, form = get_form(data)
print(storage, form); print()

dt = np.dtype([
    ("a", int),
    ("b", float),
], align=True)
arr2 = np.zeros(2, dtype=dt)
data = {"arr":  arr, "arr2": arr2, "v": arr2[0], "z": 10}
storage, form = get_form(data)
print(storage, form); print()

old_data = data
dt = np.dtype([
    ("a", int),
    ("b", object),
], align=True)
data = np.zeros(1,dtype=dt)[0]
data["a"] = 10
data["b"] = old_data
storage, form = get_form(data)
print(storage, form); print()
