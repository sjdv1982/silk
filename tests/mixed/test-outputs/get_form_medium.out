"mixed-binary" {
  "contiguous": true,
  "identical": true,
  "items": {
    "bytesize": 16,
    "order": [
      "a",
      "b"
    ],
    "properties": {
      "a": {
        "bytesize": 8,
        "storage": "pure-binary",
        "type": "integer",
        "unsigned": false
      },
      "b": {
        "identical": true,
        "items": "integer",
        "shape": [
          3
        ],
        "storage": "pure-plain",
        "type": "array"
      }
    },
    "type": "object"
  },
  "shape": [
    2
  ],
  "type": "array"
} 

"mixed-binary" {
  "contiguous": true,
  "identical": false,
  "items": [
    {
      "bytesize": 16,
      "order": [
        "a",
        "b"
      ],
      "properties": {
        "a": {
          "bytesize": 8,
          "storage": "pure-binary",
          "type": "integer",
          "unsigned": false
        },
        "b": {
          "identical": true,
          "items": "integer",
          "shape": [
            3
          ],
          "storage": "pure-plain",
          "type": "array"
        }
      },
      "type": "object"
    },
    {
      "bytesize": 16,
      "order": [
        "a",
        "b"
      ],
      "properties": {
        "a": {
          "bytesize": 8,
          "storage": "pure-binary",
          "type": "integer",
          "unsigned": false
        },
        "b": {
          "identical": true,
          "items": "number",
          "shape": [
            3
          ],
          "storage": "pure-plain",
          "type": "array"
        }
      },
      "type": "object"
    }
  ],
  "shape": [
    2
  ],
  "type": "array"
} 

"mixed-plain" {
  "identical": true,
  "items": {
    "contiguous": true,
    "identical": true,
    "items": {
      "bytesize": 8,
      "type": "number"
    },
    "shape": [
      10
    ],
    "storage": "pure-binary",
    "type": "array"
  },
  "shape": [
    3
  ],
  "type": "array"
} 

"mixed-plain" {
  "identical": false,
  "items": [
    "integer",
    {
      "contiguous": true,
      "identical": true,
      "items": {
        "bytesize": 8,
        "type": "number"
      },
      "shape": [
        10
      ],
      "storage": "pure-binary",
      "type": "array"
    },
    "null"
  ],
  "shape": [
    3
  ],
  "type": "array"
} 

"mixed-plain" {
  "properties": {
    "arr": {
      "contiguous": true,
      "identical": true,
      "items": {
        "bytesize": 8,
        "type": "number"
      },
      "shape": [
        10
      ],
      "storage": "pure-binary",
      "type": "array"
    },
    "arr2": {
      "contiguous": true,
      "identical": true,
      "items": {
        "bytesize": 16,
        "order": [
          "a",
          "b"
        ],
        "properties": {
          "a": {
            "bytesize": 8,
            "type": "integer",
            "unsigned": false
          },
          "b": {
            "bytesize": 8,
            "type": "number"
          }
        },
        "type": "object"
      },
      "shape": [
        2
      ],
      "storage": "pure-binary",
      "type": "array"
    },
    "v": {
      "bytesize": 16,
      "order": [
        "a",
        "b"
      ],
      "properties": {
        "a": {
          "bytesize": 8,
          "type": "integer",
          "unsigned": false
        },
        "b": {
          "bytesize": 8,
          "type": "number"
        }
      },
      "storage": "pure-binary",
      "type": "object"
    },
    "z": "integer"
  },
  "type": "object"
} 

"mixed-binary" {
  "bytesize": 16,
  "order": [
    "a",
    "b"
  ],
  "properties": {
    "a": {
      "bytesize": 8,
      "storage": "pure-binary",
      "type": "integer",
      "unsigned": false
    },
    "b": {
      "properties": {
        "arr": {
          "contiguous": true,
          "identical": true,
          "items": {
            "bytesize": 8,
            "type": "number"
          },
          "shape": [
            10
          ],
          "storage": "pure-binary",
          "type": "array"
        },
        "arr2": {
          "contiguous": true,
          "identical": true,
          "items": {
            "bytesize": 16,
            "order": [
              "a",
              "b"
            ],
            "properties": {
              "a": {
                "bytesize": 8,
                "type": "integer",
                "unsigned": false
              },
              "b": {
                "bytesize": 8,
                "type": "number"
              }
            },
            "type": "object"
          },
          "shape": [
            2
          ],
          "storage": "pure-binary",
          "type": "array"
        },
        "v": {
          "bytesize": 16,
          "order": [
            "a",
            "b"
          ],
          "properties": {
            "a": {
              "bytesize": 8,
              "type": "integer",
              "unsigned": false
            },
            "b": {
              "bytesize": 8,
              "type": "number"
            }
          },
          "storage": "pure-binary",
          "type": "object"
        },
        "z": "integer"
      },
      "storage": "mixed-plain",
      "type": "object"
    }
  },
  "type": "object"
} 

